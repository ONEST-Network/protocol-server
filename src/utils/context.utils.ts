import moment from "moment";
import fs from "fs";
import { v4 as uuid_v4 } from "uuid";
import { Exception, ExceptionType } from "../models/exception.model";
import { AppMode } from "../schemas/configs/app.config.schema";
import { ActionUtils } from "./actions.utils";
import { getConfig } from "./config.utils";

export const contextBuilder = async (
  context: any,
  action: string
): Promise<any> => {
  if (!context) {
    throw new Exception(
      ExceptionType.Context_NotFound,
      "Context not found",
      404
    );
  }
  if (!context.domain) {
    throw new Exception(
      ExceptionType.Context_DomainNotFound,
      "Domain not found in the context",
      404
    );
  }
  if (!context.version && !context?.core_version) {
    throw new Exception(
      ExceptionType.Context_CoreVersionNotFound,
      "Core version not found in the context",
      404
    );
  }

  const rawdata: any = await fs.promises.readFile(
    `schemas/context_${context.version?context.version:context.core_version}.json`
  );


  const defaltConfig = getConfig()
  const id=uuid_v4()

  const dynamicContext = Object.entries(JSON.parse(rawdata)).reduce(
    (accum: any, [key, val]: any) => {
      accum[key] = eval(val);
      return accum;
    },
    {
      
      ttl: moment.duration(getConfig().app.ttl, "ms").toISOString(),
      action: ActionUtils.parseAction(context.action)
     
    }
  );
 // console.log(dynamicContext)
  return dynamicContext ;
 
};

