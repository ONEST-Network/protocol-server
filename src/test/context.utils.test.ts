import { describe, expect, test } from "@jest/globals";
import { v4 as uuid_v4 } from "uuid";
import { contextBuilder } from "../utils/context.utils";
import { Exception, ExceptionType } from "../models/exception.model";
import fs from "fs";
import moment from "moment";

describe("bapContextBuilder", () => {
  it("throws an exception if context is missing", async () => {
    try {
      await contextBuilder(null, "someAction");
    } catch (error) {
      expect(error).toEqual(
        new Exception(ExceptionType.Context_NotFound, "Context not found", 404)
      );
    }
  });

  it("throws an exception if context domain is missing", async () => {
    try {
      await contextBuilder({}, "someAction");
    } catch (error) {
      expect(error).toEqual(
        new Exception(
          ExceptionType.Context_DomainNotFound,
          "Domain not found in the context",
          404
        )
      );
    }
  });
  it('throws an exception if both context version and core_version are missing', async () => {
    const context = { domain: 'someDomain', action: 'someAction' };
    await expect(contextBuilder(context, 'someAction')).rejects.toThrowErrorMatchingSnapshot();
  });

  it("builds bapContext with minimum required fields", async () => {
    const result = await contextBuilder(
      {
        domain: "mobility",
        core_version: "0.9.4",
        bap_id: "mobility@beckn.io",
        bap_uri: "mobility@beckn.io",
        country: "INDIA",
        action: "search",
        city: "CHENNAI",
        bpp_id: "becknsandbox@becknprotocol.io",
        bpp_uri: "becknsandbox@becknprotocol.io",
        transaction_id: "1234",
      },
      "search"
    );

    expect(result).toMatchObject({
      domain: "mobility",
        core_version: "0.9.4",
        bap_id: "mobility@beckn.io",
        bap_uri: "mobility@beckn.io",
        country: "INDIA",
        action: "search",
        city: "CHENNAI",
        bpp_id: "becknsandbox@becknprotocol.io",
        bpp_uri: "becknsandbox@becknprotocol.io",
        transaction_id: "1234",
      ttl: expect.any(String),
      timestamp: expect.any(String),
      message_id: expect.any(String),
    });
  });

  it("builds bapContext with minimum required fields", async () => {
    const result = await contextBuilder(
      {
        domain: "mobility",
        version: "1.1.0",
        bap_id: "mobility@beckn.io",
        bap_uri: "mobility@beckn.io",
        location: {
          city: {
              name: "Bangalore",
              code: "std:080"
          },
          country: {
              name: "India",
              code: "IND"
          }
      },
        action: "search",
        bpp_id: "becknsandbox@becknprotocol.io",
        bpp_uri: "becknsandbox@becknprotocol.io",
        transaction_id: "123",
      },
      "search"
    );

    expect(result).toMatchObject({
      domain: "mobility",
      version: "1.1.0",
      bap_id: "mobility@beckn.io",
      bap_uri: "mobility@beckn.io",
      location: {
        city: {
            name: "Bangalore",
            code: "std:080"
        },
        country: {
            name: "India",
            code: "IND"
        }
    },
      action: "search",
      bpp_id: "becknsandbox@becknprotocol.io",
      bpp_uri: "becknsandbox@becknprotocol.io",
      transaction_id: "123",
      ttl: expect.any(String),
      timestamp: expect.any(String),
      message_id: expect.any(String),
    });
  });
  it('reads context schema based on context version', async () => {
    fs.promises.readFile = jest.fn().mockResolvedValue(JSON.stringify({}));
    const result = await contextBuilder(
      {
        domain: "mobility",
        core_version: "0.9.4",
        bap_id: "mobility@beckn.io",
      bap_uri: "mobility@beckn.io",
        country: "INDIA",
        action: "search",
        city: "CHENNAI",
        bpp_id: "becknsandbox@becknprotocol.io",
      bpp_uri: "becknsandbox@becknprotocol.io",
       
        transaction_id: "123",
      },
      "search"
    );
    expect(fs.promises.readFile).toHaveBeenCalledWith(expect.stringContaining('schemas/context_0.9.4.json'));
  });

  it('reads context schema based on core_version if version is missing', async () => {
    fs.promises.readFile = jest.fn().mockResolvedValue(JSON.stringify({}));
    const result = await contextBuilder(
      {
        domain: "mobility",
        core_version: "0.9.4",
        bap_id: "mobility@beckn.io",
      bap_uri: "mobility@beckn.io",
        country: "INDIA",
        action: "search",
        city: "CHENNAI",
        bpp_id: "becknsandbox@becknprotocol.io",
      bpp_uri: "becknsandbox@becknprotocol.io",
        transaction_id: "123",
      },
      "search"
    );
    expect(fs.promises.readFile).toHaveBeenCalledWith(expect.stringContaining('schemas/context_0.9.4.json'));
  });
  
 
 
  it('uses provided context.action for action parsing', async () => {
   
    const result = await contextBuilder(
      {
        domain: "mobility",
        core_version: "0.9.4",
        bap_id: "mobility@beckn.io",
      bap_uri: "mobility@beckn.io",
        country: "INDIA",
        action: "search",
        city: "CHENNAI",
        bpp_id: "becknsandbox@becknprotocol.io",
      bpp_uri: "becknsandbox@becknprotocol.io",
        transaction_id: "123",
      },
      "search"
    );

    expect(result.action).toEqual(expect.objectContaining({}));
  });

  
   it('handles missing schema file', async () => {
    fs.promises.readFile = jest.fn().mockRejectedValue(new Error('File not found'));
    const context = {
      domain: "mobility",
      core_version: "0.9.4",
      bap_id: "mobility@beckn.io",
    bap_uri: "mobility@beckn.io",
      country: "INDIA",
      action: "search",
      city: "CHENNAI",
      bpp_id: "becknsandbox@becknprotocol.io",
    bpp_uri: "becknsandbox@becknprotocol.io",
      transaction_id: "132",
    };

    await expect(contextBuilder(context, 'someAction')).rejects.toThrowErrorMatchingSnapshot();
  });

});
