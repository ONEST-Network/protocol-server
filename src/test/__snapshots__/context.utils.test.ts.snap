// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bapContextBuilder handles missing schema file 1`] = `"File not found"`;

exports[`bapContextBuilder throws an exception if both context version and core_version are missing 1`] = `"Core version not found in the context"`;
